Major Cult Change

Xhuis opened the pull request:
The current code for runes is god-awful. Like, seriously terrible. This PR aims to fix that. Rather than having every rune effect defined in a single attack_hand proc, this PR will separate the runes into subtypes, which supports for more modular runes as well as cleans up the code.

It also separates invocation into a two different procs. The first of these will determine if the rune can be invoked; if that passes, the second will cause the rune's effects. There is also a third proc that will cause effects if the rune fails to invoke.

Research words have been removed, and all runes are now available from the start of the round. This is highly subject to change once some feedback is in from actual cult rounds.

Miscellaneous tweaks:
- Blood boil runes now explode when used
- Convert runes can now work solo, but take time
- Armor runes now only provide the basic armor set
- Tomes now do a static 15 burn damage and play a sound
- If the Escape objective is rolled, the amount of people that must escape will scale with the round population to a factor of 5 (i.e. 50 players = 10 cultists)
### This is merge-ready, please merge this as soon as you can.

Commit:
Major cult changes

Anonus commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,45 +1,133 @@
-	else
-		call(/obj/effect/rune/proc/fizzle)()
+	else if(istype(I, /obj/item/weapon/nullrod))
+		user << "<span class='notice'>You disrupt the vile magic with [I].</span>"
```
the vile magic of the [src]

Miauw62 replied:
>lmao anonus you are suggesting xhuis make his code worse.
if you add the here it will result in the the rune

Anonus commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,45 +1,133 @@
+
+/obj/effect/rune/attack_hand(mob/living/user)
+	if(!iscultist(user))
+		user << "<span class='warning'>You aren't able to understand the words of [src].</span>"
```
the [src]

Xhuis replied:
THE [src]ENING

Anonus replied:
[src]ERY!

Miauw62 replied:
lmao anonus you are suggesting xhuis make his code worse.
if you add the here it will result in the the rune

duncathan commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,45 +1,134 @@
+/obj/effect/rune
+	name = "rune"
+	desc = "A strange collection of symbols drawn in blood."
+	var/cultist_desc = "An malformed, improper rune. Invoking this may cause problems." //This is shown to cultists who examine the rune in order to determine its true purpose.
```
A malformed*

duncathan commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,45 +1,134 @@
+	layer = TURF_LAYER
+
+	var/invocation = "Aiy ele-mayo!" //This is said by cultists when the rune is invoked.
+	var/type = "invalid" //This is used to determine the icon and color of the rune after it is drawn.
```
afaik this will prevent it from compiling due to the existence of the type var already (the one every datum has)

oisin100 commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,45 +1,134 @@
+	unacidable = 1
+	layer = TURF_LAYER
+
+	var/invocation = "Aiy ele-mayo!" //This is said by cultists when the rune is invoked.
```
Aiy ale-mayo = Ayy lmao

Xhuis commented:
Did some more updates Video demo (ignore the random bwoink): https://dl.dropboxusercontent.com/u/116194515/rune_refactor_1.mp4

What you can't see here is the popup windows - Raise Dead lets you choose the corpses, for instance. I'd like feedback on the new 3x3 sprite for the Nar-Sie rune as well as accompanying sound.

Anonus commented:
Sound is eh, sprite is also eh, to be honest.

Xhuis commented:
I'm a coder, not a spriter vOv

Xhuis commented:
Updated: https://i.gyazo.com/e4819888b61116ef8f3eaeac2be81e49.png

Xhuis commented:
I added the rest of the runes with a few exceptions. Some functionalities have been expanded or removed. For instance, there is no free cultist rune, but summon cultist no longer requires the target to be free. Talismans and their respective runes will come later.

JJRcop commented on code/game/gamemodes/cult/ritual.dm:
```diff
@@ -82,762 +88,103 @@ var/engwords = list("travel", "blood", "join", "hell", "destroy", "technology",
+	user.visible_message("<span class='warning'>[user] creates a strange circle in their own blood.</span>", \
+						 "<span class='danger'>You finish drawing the arcane markings of the Geometer.</span>")
+	var/obj/effect/rune/rune_type
+	switch(rune_to_scribe)
```
I'd look into the ```typesof``` proc so you can dynamically get all runes, unless you don't want that.
(Also, if you do decide to do this, I'd cache the result because you can't add more types at runtime.)

Xhuis replied:
I figured that wouldn't work because it would display the object path, but I'll look into it.

RemieRichards replied:
```
var/list/runes = list()
for(var/t in typesof(/obj/effect/rune))
    var/obj/effect/rune/r = t
    runes[initial(r.name)] = t
```

JJRcop replied:
Except of course in this case it would be ``initial(r.cultist_name)``

RemieRichards commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,518 @@
+	layer = TURF_LAYER
+	color = rgb(255,0,0)
+
+	var/invocation = "Aiy ele-mayo!" //This is said by cultists when the rune is invoked.
```
that's now even how you pronounce it
it's "eye-la-mao"

correct it or pick a better joke or something else.

Miauw62 replied:
this is the base rune subtype, this invocation should never happen anyway so i dont see the issue.

RemieRichards replied:
It's a joke, lighten up.

RemieRichards commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,518 @@
 
-///EMP: Lets out a large EMP blast
+There are a few different procs each rune runs through when a cultist activates it.
+begin_invoke() is called when a cultist activates the rune with an empty hand. If there are multiple cultists, this rune determines if the required amount is nearby.
```
it's ```can_invoke()``` now

RemieRichards commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,518 @@
+	var/obj/effect/rune/selected_rune = pick(potential_runes)
+	user.visible_message("<span class='warning'>[user] vanishes in a flash of red light!</span>", \
+						 "<span class='danger'>Your vision blurs, and you suddenly appear somewhere else.</span>")
+	user.loc = get_turf(selected_rune)
```
```user.Move(get_turf(selected_rune))``` or ```user.forceMove(get_turf(selected_rune))``` (I believe both handle everything)
directly setting loc will fail to trigger area based stuff and other junk.

RemieRichards commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,518 @@
+	visible_message("<span class='warning'>[src] flares brightly, then slowly dulls and appears mundane.</span>")
+	for(var/obj/effect/rune/R in orange(3,src))
+		R.desc = "A rune drawn in crayon."
+	desc = "A rune drawn in crayon."
```
... just make it ```range()``` and not ```orange()```

RemieRichards commented:
Gotta go, will finish review later.

Xhuis commented:
Got the tome working. It lets you create runes of every type (I have yet to do names and removing the malformed type from writable ones) and you spawn with a tome now rather than a supply talisman. I have all the basic runes in and working - unfortunately the Rite of Binding seems to have an issue where the paper used for the talisman isn't deleted.

Also, it compiles! \o/

RemieRichards commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,649 @@
-		user.equip_to_slot_or_del(new /obj/item/clothing/suit/magusred(user), slot_wear_suit)
-		user.equip_to_slot_or_del(new /obj/item/clothing/shoes/cult(user), slot_shoes)
-		user.spellremove(user)
-		usr.visible_mes
```
turn the path of the rune into a string, split it by ```/```'s find the last word (including ```_```, splitting by ```/``` will keep ```_``` intact which is good) then append this word on to the end of a ```"/obj/item/weapon/paper/talisman/``` string, so you have the path of the talisman, then turn the string into a path, and have it spit out a malformed if it's not a path. you'll need ```path2text()``` ```text2path()``` and ```ispath()```

Xhuis replied:
path2text() doesn't seem to be a proc.

RemieRichards replied:
Haha, yeah uh, path2text() doesn't exist because it's simple and I was tired when I wrote that bit.
path2text() is basically just ```"[thing.type]"```

"Rune2Talisman"
```
var/Rtype = "[R.type]"
var/list/Rsplit = text2list(Rtype,"/")
var/RuneType = Rsplit[4]
var/TailismanType = "/obj/item/weapon/paper/talisman/[RuneType]"
if(ispath(TalisimanType))
    new TalismanType(get_turf(src))
else
    new /obj/item/weapon/paper/talisman/malformed (get_turf(src))
```

RemieRichards commented on code/game/gamemodes/cult/talisman.dm:
```diff
@@ -1,125 +1,199 @@
+	if(!rune_to_imbue || !istype(rune_to_imbue))
+		return
+	var/talisman_type
+	switch(rune_to_imbue.type)
```
See above

Xhuis commented:
Cleaned up the talisman rune as per feedback. It's having an issue where it won't actually create a talisman or do anything with the paper - it just deletes both runes. Using debug code I know that I have the right idea - it even worked before I added in the qdel()s.

RemieRichards commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,635 @@
-		user.equip_to_slot_or_del(new /obj/item/clothing/suit/magusred(user), slot_wear_suit)
-		user.equip_to_slot_or_del(new /obj/item/clothing/shoes/cult(user), slot_shoes)
-		user.spellremove(user)
-		usr.visible_message("\red The rune disappears with a
```
p.sure this needs to be the 4th index, not the 5th.

Xhuis replied:
It actually didn't work when I used [4]. [5] is the only way I got it to work.

RemieRichards replied:
unless I'm mistaken, ```/obj/effect/rune/WORD``` should only produce a 4 element list when ```text2list(x,"/")```'d I guess it's counting some extra space at the beginning, which is stupid.

Xhuis commented:
All right, this binding rune is giving me loads of grief and I need help.
Here's the code as it is on my local copy: http://hastebin.com/ebimatowap

The intended functionality is that the binding rune creates a talisman out of a paper and a nearby rune, deleting both, creating the talisman object on the rune's turf, and then leaving the binding rune. Instead, both runes are deleted and the original paper remains.

Does anyone know why this might be happening?

Xhuis commented:
I've started including cult structures. They're going to be required for a lot of different runes, and can have different effects. Altars can now be used to sacrifice people to the Geometer, and a lot of grammar has been changed.

Shapsy commented on code/game/gamemodes/cult/cult_items.dm:
```diff
@@ -86,18 +100,18 @@
 	icon_state = "cult_armor"
 	item_state = "cult_armor"
-	desc = "A bulky suit of armor, bristling with spikes. It looks space proof."
+	desc = "A heavily-armored piece of armor worn by warriors of the Nar-Sian cult. It can withstand hard vacuum."
```
Heavily armored piece of armor?

Xhuis replied:
sssh

Xhuis commented:
Squashed commits. I've leaving a small bit of the framework in for new stuff, but for now I want to get the actual refactor out of the way.

Fox-McCloud commented:
Just a note: this removes the conversion of holy water to unholy water--thus making mass chem bombings with holy water an uncounterable strategy for use against cultists.

Xhuis commented:
I'll be fixing it, don't worry - this is still WIP

Aloraydrel commented:
Its happening1

Aloraydrel commented:
Its happening!

fleure commented:
Technically this isn't a refactor since it changes functionality/behavior
...
I'll leave now.

RemieRichards commented:
The old robes are ass, they were replaced for a reason, don't go back to them
:-1: :x: :100: 

Summon sound is ass
:-1: :x: :100:

Xhuis commented:
Summon sound is better than hallucination sound, but is being reduced in volume.
Cult robes are being replaced with old versions again.

Xhuis commented:
Lowered the Nar-Sie summon sound volume, this is the new sound.
https://dl.dropboxusercontent.com/u/116194515/narsie_sound.mp4
Re-added supply talismans.
Changed cult robe sprites back to normal and updates names and descriptions of several items.

Metacide commented:
New sound is a lot better, this all looks very impressive. 

 :+1:

Shadowlight213 commented:
Wish you could keep the "I'm here" in the summon sound. Maybe change how it sounds but keep the words? I'm here has become sort of the classic narnar sound.

Xhuis commented:
I'm putting this on hiatus until people decide on whether they want to replace research words and, if so, a suitable replacement.

Xhuis commented:
I've added grammar-based runes. Check OP for more.

Xhuis commented:
This can be considered merge-ready!

RemieRichards commented on code/game/gamemodes/cult/ritual.dm:
```diff
@@ -82,762 +86,207 @@ var/engwords = list("travel", "blood", "join", "hell", "destroy", "technology",
+	if(!entered_grammar)
+		return
+	for(var/T in typesof(/obj/effect/rune))
+		var/obj/effect/rune/R = new T(src)
```
NOT HAPPENING.
Learn to use ```intial()``` to obtain variables from typepaths, creating instances from ```typesof()``` in loops solely to check one variable is not allowed.

Xhuis replied:
This is literally the only way I could get it to work after hours of debugging. If I don't go with this, it's going to take me a long time to fix it myself unless you can come up with a solution.

RemieRichards replied:
```
for(var/T in typesof(/obj/effect/rune))
    var/obj/effect/rune/R = T
    if(initial(R.grammar) == entered_grammar)
        rune_to_scribe = T
        break //I added this for performance improvements to your code, it's nothing to do with initial()
```

Xhuis replied:
I'll plug this in and see how it works.

Xhuis replied:
I'm having the same problem as I did before I changed it to that.

optimumtact replied:
Remie I think a global list of rune words -> rune type is a far better solution than intial or any other type based magic.

This method using typesof is unnecessary black magic and will confuse future devs imo.

I actually helped Xhuis with this and creating an instance of the rune type was the only way I could actually get the grammar to be read without a runtime.

RemieRichards replied:
@optimumtact it's not black magic, it's used multiple times in the codebase (Changeling Team Objectives, something MSO did recently, some other places, etc.)
we're not ```newing``` every instance from a ```typesof()```, it's -just- not happening, especially since ```initial()``` has a feature SPECIFICALLY DESIGNED to allow us to get around that.

An assoc list of words to rune type would be good yes, provided it's constructed at roundstart from the actual runes so that if rune words change the code still works.

But no, ```initial()``` is not black magic.
As for the snippit it above, it should work, as it's identical to a bunch of initial()-on-typepath stuff I've written, Xhuis probably didn't implement it right or is having a different error he's misappropriating as being this

optimumtact replied:
I'm not arguing for the newing of the instances, it's obvious to me that it's not a nice way to do it, it's just the only way I could get it to work with the typesof method, I guess I did something wrong, or xhuis mistyped one of my code lines.

I'm not sure why a compile time assoc list is such a huge deal though - sure you get some ease of use for future coders with runtime generated lists, but IMO it's not worth having to write code like initial that will confuse newbies needlessly.

Anyway I'm kinda bikeshedding off the original point and I don't have strong opinions on this, so if we continue that debate, better to do it off this issue and on forums/irc or something.

JJRcop replied:
I'm gonna slip in that instead of roundstart maybe do it the first time it's needed.

RemieRichards commented on code/game/gamemodes/cult/ritual.dm:
```diff
@@ -82,762 +86,207 @@ var/engwords = list("travel", "blood", "join", "hell", "destroy", "technology",
+			qdel(R)
+	if(!rune_to_scribe)
+		rune_to_scribe = /obj/effect/rune/malformed //If improper grammar is entered, the rune becomes malformed
+	if(rune_to_scribe == /obj/effect/rune/teleport)
```
change this to a check for ```needs_keyword``` or some such, this way we can easily add more runes that use keywords in the future (tl;dr this bit is snowflake and no better than the old rune code)

Xhuis replied:
Teleport is the only rune that uses a keyword. If I tried to check for rune_to_scribe.keyword it would give me a compile error since the keyword variable doesn't exist. The only way it would work is if I typecasted rune_to_scribe is a teleport rune, which is bulkier and no different than what's in place already.

RemieRichards replied:
That's not even remotely related to what I said.
you move keyword up to ```/rune```, and add a ```req_keyword``` var, done, simple as, NO SNOWFLAKE.

Xhuis replied:
All right, I'll try, I guess. vOv

Xhuis commented:
I got rune creation and keywords majorly optimized - huge thanks to @RemieRichards - and I just need to wrap up a few more things before this will be merge-ready. If I've forgotten anything that's originally in Cult, let me know and I'll make sure it gets in.

RemieRichards commented:
ding ding ding we have a weeeeiner.

Fox-McCloud commented:
Don't forget unholy water and removing it with tomes.

Xhuis commented:
I have removed all instances of research words as well as finished up some extra stuff. Once I re-add the Teleport Other rune, this can finally be considered merge-ready.

Xhuis commented:
I fixed up the code a bit and re-added the Teleport Other rune. This has been pretty extensively tested, and I can safely say that it's probably merge-ready, as long as you find the code up-to-date and with no problems.

Cheridan commented:
>This does mean, however, that people who memorize the combinations can create all the runes from the beginning.

If you're going to remove research, actually remove it and just have people pick runes from a list. Right now the system is just going to confuse people who don't realize they can just copy-paste rune words from the wiki to get whatever rune they want.

RemieRichards commented:
@Xhuis do something along @Cheridan 's line of inquiry, and I'll have a look at the PR as a whole when I get back in 4/5 hours.

Xhuis commented:
All right, I left the framework in for grammar in case I decide to rework it, but for now cultists can choose from all the available runes when they scribe one.

Razharas commented:
I like research so much
Its the best part about cult, fuck all kinds of conversions and narsie and shit

Commit:
Conflicts 1/3

Commit:
Conflicts 2/3

Commit:
Conflicts 3/3

Xhuis commented:
@Razharas stop being a cuck

Conflicts fixed, should be merge-ready

Kromgar commented:
Thing is this might actually make cult viable if we start off with all the runes.

Anonus commented:
I'm not sure that's a bad thing, given that cult so far has been a rollercoaster of viability.
Also this probably isn't WIP anymore.

Xhuis commented:
It isn't WIP, it's merge-ready

Razharas commented:
Wait so it is not per-round randomized word combinations there?
If you can just learn them from wiki why have any sac revealing and shit if its static info?

Xhuis commented:
I forgot to update the OP - for now cultists merely have all the runes available and pick from a list. Will update the OP when I get home.

Xhuis commented:
Do maintainers have any major issues with merging this?

Xhuis commented:
merge please

Commit:
Conflicts 1/2

Commit:
Conflicts 2/2

Commit:
Conflicts 1/3

Commit:
Conflicts 2/3

Commit:
Conflicts 3/3

Xhuis commented:
Conflicts are fixed. I'd appreciate if this gets merged, unless maintainers have some major issue with this.

Commit:
Conflicts 1/?

Commit:
Merge remote-tracking branch 'remotes/upstream/master' into cult_update

Commit:
Conflicts 3/3

Xhuis commented:
Merge conflicts should be fixed, again.

JJRcop commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,661 +1,628 @@
+	qdel(src)
 
 
+<<<<<<< HEAD
```
you didn't actually resolve conflicts

Commit:
Fixes an errant snippet of code

Anonus commented:
Uh, you fucked up the astral communion rune, the proc's gone

Xhuis commented:
# why couldn't you have just merged it when there was no conflicts

JJRcop commented:
Not a good idea to just remove things when there's a conflict, you should pick through it by examining what it does and either reimplement it or simply move it or whatever is necessary.

Also I recommend saying in the OP that it's ready to merge, but that you will resolve conflicts only when a collaborator is prepared to merge it shortly after, that's what I've done a few times and it worked out good.

Commit:
Fixes everything

Xhuis commented:
### it's ready boys

RemieRichards commented:
@Xhuis *ahem*

RemieRichards commented on code/game/gamemodes/cult/cult_items.dm:
```diff
@@ -41,18 +44,9 @@
 	max_heat_protection_temperature = HELMET_MAX_TEMP_PROTECT
 
-
-/obj/item/clothing/head/culthood/alt
```
Didn't you revert back to using the proper cult sprites? (which confusingly enough, is defined here as the "alt" ones)

RemieRichards commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,982 @@
+	if(iscultist(user))
+		user << "<b>Name:</b> [cultist_name]"
+		user << "<b>Effects:</b> [cultist_desc]"
+		//user << "<b>Words:</b> [grammar]"
```
this would be useful, why not return it?

Xhuis replied:
The grammar isn't used right now for anything - I left it in as framework

RemieRichards commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,982 @@
-
-
-///EMP: Lets out a large EMP blast
+		target = targets[targets.len]
```
did you mean ```target = targets[target]```?
because right now you're disregarding what the user chose, and always picking the last person in the list.

Xhuis replied:
Definitely, I'll fix that

RemieRichards replied:
Yeah this diff looks different to when I placed that comment, github's wack.

RemieRichards commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,982 @@
+		if(!iscultist(M))
+			possible_targets.Add(M)
+	var/mob/offering
+	if(possible_targets.len < 1)
```
if there's no targets... make the user pick from the empty list?

Xhuis replied:
Misplaced sign, will fix

RemieRichards commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,982 @@
+	grammar = "mgar jatkaa karazet"
+
+/obj/effect/rune/blind/invoke(mob/living/user)
+	visible_message("<span class='warning'>[src] emits a blinding red flash!</span>
```
using ```.loc``` ever. cmon on, ```get_turf()```

Xhuis replied:
There is no .loc here.

RemieRichards replied:
the diff's moved.

RemieRichards commented:
Right ok, it seems fine besides points I've raised.
most of your selection of targets is very strange, sort of ass backwards.

Commit:
Fixes a misplaced sign

RemieRichards commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,982 @@
+	grammar = "mgar jatkaa karazet"
+
+/obj/effect/rune/blind/invoke(mob/living/user)
+	visible_message("<span class='warning'>[src] emits a blinding red flash!</span>
```
github pls comment on the right line this time.

RemieRichards replied:
FUCK OFF WITH THIS SHIT GITHUB.

RemieRichards replied:
Just fix all the .locs to be checking against the turf.

Xhuis replied:
I can't, I am blind without GitHub :^)

Commit:
Replaces most instances of .loc with get_turf()

Supermichael777 commented:
> sac rune only takes one person
holy fuck no.

Xhuis commented:
I forgot to remove that. It takes three to sacrifice things.

Xhuis commented:
Many-daily reminder that this is merge-ready.
@Miauw62 @RemieRichards @Dorsisdwarf @Jordie0608

phil235 commented on html/changelogs/Xhuis-CultRefactor.yml:
```diff
@@ -0,0 +1,12 @@
+changes: 
+  - experimental: "Cultists will now be able to create all runes without needing to research or collect words. This is subject to change."
+  - tweak: "Cult code has been significantly improved and will run much faster and be more responsive."
+  - tweak: "Many runes have received tweaks or changes. Conversion runes can now be invoked with less than three cultists (but take some time), sacrifice runes only require one cultist, armor runes no longer give different armors, astral journey runes now pull the user back if they're moved and give them a different color, summon and freeing runes have been combined and require two cultists rathe ...
```
`sacrifice runes only require one cultist` 
You forgot to remove that part.

Dorsisdwarf commented:
Don't ping me on mergebegging please, I'm still not allowed to merge commits.

phil235 commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,983 @@
+	var/turf/T = get_turf(src)
+	var/list/possible_targets = list()
+	for(var/mob/living/M in T.contents)
+		if(!iscultist(M))
```
You need to also check that the cultist can participate in the invocation, i.e. !M.stat , or even !M.incapacitated() if you're okay with not letting cuffed cultist do invocations.

phil235 commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,983 @@
-	return
+		var/cultists_in_range = 0
+		for(var/mob/living/L in orange(1, src))
+			if(iscultist(L))
```
Same here, need an !L.stat

Xhuis replied:
Will fix all these, but I'll leave the ability to invoke if weakened because the cultists nearby are chanting. Only the invoker is actually touching the rune.

Xhuis replied:
Fixed

phil235 commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,983 @@
+	grammar = "mgar jatkaa karazet"
+
+/obj/effect/rune/blind/invoke(mob/living/user)
+	visible_message("<span class='warning'>[src] emits a blinding red flash!</span>
```
Plasteel sheets, not metal ones.

phil235 replied:
This doesn't cover the case of having multiple stacks of plasteel, each with less than 5 sheets, but with more than 5 in total.

phil235 replied:
The runes have no click cooldown, which means they can be used to spam incantation messages. It also makes the leeching rune pretty OP, since you can kill someone with it in a matter of seconds.

Xhuis replied:
This has always been the case. If it's a problem, I'll add a one-second click delay.

Xhuis replied:
I think GitHub's line commenting is being strange. Where is this error?

phil235 commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,983 @@
+		for(var/mob/living/M in orange(1,src))
+			if(iscultist(M) && M != user)
+				cultists_nearby++
+				M.say(invocation)
```
do the cultists need to say the invocation twice? (in can_invoke() and here)

Xhuis replied:
The Raise Dead rune doesn't have a default invocation because of the reference I kept in.

phil235 replied:
This isn't the raise dead rune, it's the sacrifice rune.

Xhuis replied:
Odd. Diff viewer is telling me it's the Raise Dead rune. Anyway, this is intentional - the rune only requires one cultist, but it searches around for three if the target meets certain requirements.

phil235 commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,983 @@
+			if(iscultist(M) && M != user)
+				cultists_nearby++
+				M.say(invocation)
+		if(cultists_nearby < 3)
```
This can't use cultist_required in can_invoke() ?

Xhuis replied:
The sacrifice rune requires three cultists if the target is human or robotic, but only one if it's a simple animal or monkey.

phil235 commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,983 @@
+		if(T.mind)
+			var/obj/item/device/soulstone/stone = new /obj/item/device/soulstone(get_turf(src))
+			if(!stone.transfer_soul("FORCE", T, usr)) //If it cannot be added
+				qdel(stone)
```
there's a sleep in transfer_soul(), which means you can spamclick the rune to make appear multiples shards (which will be destroyed...probably). And this can cause a runtime since we don't check if(T) again after the transfer_soul sleep.

Xhuis replied:
Fixed

phil235 commented:
Also, can you make the runes (except the big one) be clickable even if you don't click specifically on a colored line of the rune. It's really a pain to click otherwise.

Xhuis commented:
Yeah, I'll give them mouse opacity.

phil235 commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,983 @@
+		user.say("Pasnar val'keriam usinar. Savrae ines amutan. Yam'toth remium il'tarat!")
+	mob_to_sacrifice.visible_message("<span class='warning'><b>[mob_to_sacrifice] disintegrates into black mist. The mist lingers, and then flows into [mob_to_revive]!</span>")
+	mob_to_sacrifice.dust()
+	sleep(20)
```
mob_to_revive can get gibbed duringe the sleep(20), you need to check if(mob_to_revive).

Xhuis replied:
Fixed

Commit:
More work

Commit:
Adds a sanity check in the Raise Dead rune

Commit:
Adds a sanity check in the Sacrifice rune

Xhuis commented:
That should be all the issues fixed.

optimumtact commented:
thanks for your hard work on this, it's really good :+1:

phil235 commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,1005 @@
+	mob_to_revive << "<span class='cult'>\"PASNAR SAVRAE YAM'TOTH. Arise.\"</span>"
+	mob_to_revive.visible_message("<span class='warning'>[mob_to_revive] draws in a huge breath, red light shining from their eyes.</span>", \
+								  "<span class='userdanger'>You awaken suddenly from the void. You're alive!</span>")
+	mob_to_revive.revive() //This does remove disabilities and such, but the rune might actually see some use because of it!
```
The potential runtime is still here, null.stat , null.visible_message() , null.revive()

Xhuis replied:
I forgot to add a return on the check, let me fix that quickly.

Commit:
Adds a return to the Raise Dead rune

phil235 commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,1006 @@
+			if(!stone.transfer_soul("FORCE", T, usr)) //If it cannot be added
+				qdel(stone)
+			if(!T)
+				return
```
Need a `rune_in_use = 0` before returning.

phil235 commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,1006 @@
+						 "<span class='danger'>You see what lies beyond. All is revealed. While this is a wondrous experience, your physical form will waste away in this state. Hurry...</span>")
+	user.ghostize(1)
+	while(user)
+		sleep(10)
```
user can become null during this sleep(10), causing runtimes below.

phil235 commented on code/game/gamemodes/cult/runes.dm:
```diff
@@ -1,1053 +1,1006 @@
+			var/mob/dead/observer/G = user.get_ghost()
+			if(G)
+				G << "<span class='warning'><b>You suddenly feel your physical form pass on. [src]'s exertion has killed you!</b></span>"
 			return
```
Probably need a `rune_in_use = 0` after the while, in case the loop stops b/c user became null.

Xhuis replied:
Added

Commit:
Adds some more sanity checks

phil235 commented:
I'm closing this PR and will merge a copy of it with just one commit.

Fox-McCloud commented:
I don't think I've ever seen a PR be nitpicked so much =p

Not saying it's not necessary, but geezy.
